File: Python/Quick Sort.py
def swap(array, i, j) :

    temp = array[i]
    array[i] = array[j]
    array[j] = temp

def Partiton(array, left, right) :
    end = array[right]
    start = left-1

    for j in range(left, right) :
        if array[j] <= end :
            start+=1
            swap(array,start,j)

    swap(array,start+1,right)
    return (start+1)

def quicksort(array, count) :
    startIndex = 0
    endIndex = count -1
    top = -1

    stack =[]
    stack.append(startIndex)
    top +=1
    stack.append(endIndex)
    top +=1

    while top >= 0 :
        endIndex = stack.pop()
        top -= 1
        startIndex = stack.pop()
        top -= 1

        p = Partiton(array, startIndex, endIndex)

        if p -1 > startIndex :
            stack.append(startIndex)
            top += 1
            stack.append(p-1)
            top += 1

        if p + 1 < endIndex :
            stack.append(p+1)
            top += 1
            stack.append(endIndex)
            top += 1

list = [6,3,11,9,12,2,8,15,18,7,14]
print(list)

quicksort(list, len(list))

print(list)

File: C/Quick Sort.c
#include <stdio.h>
#include <stdlib.h>

void swap(int* a, int* b) {
	int tmp = *a;
	*a = *b;
	*b = tmp;
}

void quick_sort(int* array, int start, int end) {

	if (start >= end) return;

	int mid = (start + end) / 2;
	int pivot = array[mid];

	swap(&array[start], &array[mid]);

	int p = start + 1, q = end;

	while (1) {
		while (array[p] <= pivot) { p++; }
		while (array[q]>pivot) { q--; }

		if (p>q) break;

		swap(&array[p], &array[q]);
	}

	swap(&array[start], &array[q]);

	quick_sort(array, start, q - 1);
	quick_sort(array, q + 1, end);

}

int main(void) {

	int array[] = { 6,3,11,9,12,2,8,15,18,7,14 };
	int i;

	quick_sort(array, 0, sizeof(array) / sizeof(int) - 1);

	for (i = 0; i< sizeof(array)/ sizeof(int); i++) {
		printf("%4d\n", array[i]);
	}

	return 0;
}


